File: 0735-asteroid-collision/0735-asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        
        Stack<Integer> st = new Stack<>();
        
        for(int a: asteroids){
            
            while(!st.isEmpty() && (a<0 && st.peek()>0)){
                
                int sum = st.peek() + a;
                
                if(sum < 0){
                    st.pop();
                }
                else if(sum>0){
                    a=0;
                }
                else{
                    st.pop();
                    a=0;
                }
            }
            
            if(a!=0)
                st.push(a);
        }
        
        int[] res = new int[st.size()];
        int i = st.size()-1;
        
        while(!st.isEmpty()){
            res[i] = st.pop();
            i--;
        }
        
        return res;
    }
}

File: 0735-asteroid-collision/0735-asteroid-collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        
        vector<int> res;
        
        for(int a: asteroids){
            
            while(!res.empty()&&(a<0 && res.back()>0)){
                
                int sum = res.back() + a;
                
                if(sum<0){
                    res.pop_back();
                }
                else if(sum>0){
                    a=0;
                }
                else{
                    res.pop_back();
                    a=0;
                }
            }
            
            if(a!=0){
                res.push_back(a);
            }
        }
        
        return res;

    }
};